Better extensions for sum/average, deprecate others

kiwisincebirth opened the pull request:
fix #28 

Added missing Kotlin wrappers for all sum() and average() methods provided in reactor.math.MathFlux to the library

 - Added unit test to cover all of these methods using the existing unit tests as a guide
 - Include Java Doc for all new methods with @ since based on the current snapshot version. 
 - Note Inconsistency, since the existing methods have @ since based on reactor-extra version numbering.

simonbasle commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -33,6 +34,45 @@ import java.util.function.Function
+ * @author Mark Pruden
+ * @since 1.1.5
+ */
+fun <T: Number> Flux<T>.sumInt(): Mono<Int> = MathFlux.sumInt(this)
```
would there be compiler ambiguity if we kept `<T:Number> sum` above AND turn these additional methods into `<T:Int>` `<T:Long>` ?
my goal was to use Kotlin as much as possible to avoid the "dumb" style of the Java `MathFlux` (which repeats the type in the method name).

kiwisincebirth replied:
Totally get where you are coming from. Your suggestion Does create a compiler Conflict with the existing `sum(): Mono<Long>` method

However there is potentially a workable solution which achieves the outcome. But it requires replacing the existing `sum(): Mono<Long>` method with the below method.

```
inline fun <reified T: Number> Flux<T>.sum(): Mono<T> =
    when (T::class) {
        BigDecimal::class -> MathFlux.sumBigDecimal(this) as Mono<T>
        Double::class -> MathFlux.sumDouble(this) as Mono<T>
        Long::class -> MathFlux.sumLong(this) as Mono<T>
        // todo above cases need to be exhaustive
        else -> Mono.empty() // or error()
    }
```

Disadvantages 
1) Existing users of sum() may be broken since new function returns Mono<T> (not Mono<Long>) - so existing code would would need to explicitly migrate to sumLong() - OR NOT - depending on use of resultant Mono<T>

Advantages
1) The sum() method would work for all cases (probably majority) where the input types holds enough precision, and there is no expectation of type conversion.
2) The new sum() would always return the same type that was passed in so there is no ambiguity. eg. `fluxOfInts.sum()` - makes no statement about output type so user can assume same type.
3) If the user needs to increase precision or change output type then the full method e.g. e.g. `fluxOfInts.sumLong()` - makes an Explicit statement about the (output) type of result that is needed.

simonbasle replied:
I very much like that 🤩

Perhaps the same style could be applied to the map-and-sum/average case ?
And there, since the mapping is explicit, no need for individual per-type methods at all!

Regarding the disadvantage, even if I'd argue that it is more desirable for the `sum` method to maintain the type, it is indeed a breaking compatibility change 😞

can we get creative with the name and deprecate the old `sum()` perhaps?

in Reactor we have the `collect()` and `reduce()` operators.
So perhaps something like `collectSum()`/`collectAverage()`?

For the map case, `mapAndCollectSum()` and `mapAndCollectAverage()`?

Any good candidates for names from the Kotlin world?

kiwisincebirth replied:
> Perhaps the same style could be applied to the map-and-sum/average case ?
Absolutely that was the intent.

> can we get creative with the name and deprecate the old sum() perhaps?
or make the breaking change in a major version release.

> can we get creative with the name and deprecate the old sum() perhaps?
Deprecation would be a good idea in any case.

> So perhaps something like collectSum()/collectAverage()?
Fo these methods I dont have any better idea's than what you suggest

> Any good candidates for names from the Kotlin world?
In kotlin world found the sumOf() method in sequence's where a mapping function exists, suggest this would be reasonable alternative, and is shorter.

https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.sequences/sum-of.html

thus
```
Flux<String> fluxOfSentences;
Mono<Double> numberOfCharsDouble = fluxOfSentences.sumOf(s -> (double) s.length());
Mono<Int> numberOfChars = fluxOfSentences.sumOf(String::length)
Mono<Int> averageWordsPerSentence = fluxOfSentences.averageOf(s -> s.split(" ").length);
```

As for the non-mapping functions, suggest go with your suggestion until future major release where the deprecared sum() average() functions potentially could be replaced.

simonbasle replied:
Ah, sumAll might be a good one too

simonbasle commented:
Sounds like a plan! collectX, xOf and deprecation

kiwisincebirth commented:
> Sounds like a plan! collectX, xOf and deprecation

> Ah, sumAll might be a good one too

Can you clarify your preference, based on comments from above ? 

(a) collectX() and xOf()
(b) xAll() and xOf()

PS. Have committed the new methods with deprecations, except naming is not correct at this time. I was more exhaustive in the switch case statements, to account for the most common data types to avoid runtime errors.

simonbasle commented:
> (a) collectX() and xOf()
> (b) xAll() and xOf()
> (c) xAll() - Noting there is no method signature conflict between mapped and unmapped functions, so names can be identical. we can just have sumAll() and averageAll()
> (d) something else

hehe, so much choice... actually, you have a good point with (c). `sumAll()` is merely a simplified case of `sumAll(Function<T, R:Number>)` where `T` is already a `Number` sub-class.

to summarize:

 - `sumAll()` and `averageAll()` would maintain the original `T:Number` type
 - `sumAll(T -> N:Number)` and `averageAll(T -> N:Number)` would explicitly convert from any type `T` to `N` Number sub-type (so no ambiguity in the output type here)
 - `sum()`/`average()`/`sum(Function)`/`average(Function)` get deprecated

Do we still need `sumLong...sumBigDecimal` and all ?

kiwisincebirth commented:
Have made the changes to the code, specifically to name methods as sumAll() and averageAll()

> Do we still need sumLong...sumBigDecimal and all ?

I believe so - consider the following code

```
    val fluxOfInts: Flux<Int> = Flux.just(1,2,3,4,5)
    val sumLong = fluxOfInts.sumLong()
    val avgDouble = fluxOfInts.averageDouble()
```

if you remove the typed functions the above code would have to be written as

```
    val fluxOfInts: Flux<Int> = Flux.just(1,2,3,4,5)
    val sumLong = fluxOfInts.sumAll(kotlin.Int::toLong)
    val avgDouble = fluxOfInts.averageAll(kotlin.Int::toDouble)
```

Which I think is good reason to keep the typed functions.

simonbasle commented:
> Which I think is good reason to keep the typed functions.

One last naming nitpick/proposal then: what do you think about `sumAsXxx` to better convey that a conversion is potentially in play? (`sumAsLong`, `sumAsDouble`, etc...)

simonbasle commented:
oh and while I agree the `sumType()` methods have value, I really believe the `sumType(Function)` and `averageType(Function)` ones should be removed (since `sumAll(Function)`/`averageAll(Function)` would be expressive and concise enough to cover these cases).

simonbasle commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -101,10 +348,72 @@ fun <T> Flux<T>.sum(mapper: (T) -> Number): Mono<Long>
  * @param mapper a lambda converting values to [Number]
  * @author Simon Baslé
- * @since 3.1.1
+ * @since 1.0.0
```
this one should actually be deprecated like the others (since it pre-dates the PR)

kiwisincebirth replied:
all original methods are now deprecated

simonbasle commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -101,10 +348,72 @@ fun <T> Flux<T>.sum(mapper: (T) -> Number): Mono<Long>
+ * The resultant [Mono] will have the same [Number] type as the output of the mapping function
+ *
+ * If the result type (or precision) needs to be changed then a specific averageType() method
+ * should be used in preference to this method. e.g. averageDouble(mapper) can be used to compute
```
as in `sumAll(mapper)`, this shouldn't reference `averageDouble(mapper)` but simply `averageDouble()`

kiwisincebirth replied:
resolved

simonbasle commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -101,10 +348,72 @@ fun <T> Flux<T>.sum(mapper: (T) -> Number): Mono<Long>
+        BigInteger::class -> MathFlux.averageBigInteger(this, Function(mapper)) as Mono<R>
+        Double::class -> MathFlux.averageDouble(this, Function(mapper)) as Mono<R>
+        Float::class -> MathFlux.averageFloat(this, Function(mapper)) as Mono<R>
+        Long::class -> MathFlux.averageBigInteger(this, Function(mapper))
```
I assume this is for better precision of the _average_, an issue that doesn't occur in sums?

kiwisincebirth replied:
Good pickup.

It is actually because the underlying Library does not support the averaging of integer type numbers, However for some reason BigInteger is supported for average.

Is this an omission - sort of IMHO - While most people think of averages and decimal numbers (obvious reasons), an average expressed as an integer is still an average (just with less precision) 

For completeness am using Big integer to do the averages then map the result back to the requested type.

simonbasle commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -115,11 +424,66 @@ fun <T> Flux<T>.sumDouble(mapper: (T) -> Number): Mono<Double>
+ * @author Mark Pruden
+ * @since 1.1.5
+ */
+fun <T> Flux<T>.averageFloat(mapper: (T) -> Number): Mono<Float>
```
as for sumTypes(mapper), averageTypes(mapper) should not be included in this PR

kiwisincebirth replied:
Have removed these methods.

simonbasle reviewed (changes requested):
refining the javadoc, missing deprecated annotations (with replaceWith) and commented on the first of each typed mapper-based variant that these shouldn't be included. we're almost there @kiwisincebirth this is looking better and better 👍

kiwisincebirth commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -54,29 +177,106 @@ fun <T: Number> Flux<T>.sumDouble(): Mono<Double> = MathFlux.sumDouble(this)
  * @author Simon Baslé
- * @since 3.1.1
+ * @since 1.0.0
+ * @deprecated Please use averageDouble() as a direct replacement of consider more general purpose averageAll()
```
Resolved in commit

kiwisincebirth commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -86,10 +286,57 @@ fun <T: Comparable<T>> Flux<T>.max(): Mono<T> = MathFlux.max(this)
  * @author Simon Baslé
- * @since 3.1.1
+ * @since 1.0.0
+ * @deprecated Please use sumLong() as a direct replacement of consider more general purpose sumAll()
```
addressed in commit.

kiwisincebirth commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -54,29 +177,106 @@ fun <T: Number> Flux<T>.sumDouble(): Mono<Double> = MathFlux.sumDouble(this)
+ * @since 1.0.0
+ * @deprecated Please use averageDouble() as a direct replacement of consider more general purpose averageAll()
  */
 fun <T: Number> Flux<T>.average(): Mono<Double> = MathFlux.averageDouble(this)
```
Added missing annotations

kiwisincebirth commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -29,10 +56,50 @@ import java.util.function.Function
  * @author Simon Baslé
- * @since 3.1.1
+ * @since 1.0.0
+ * @deprecated Please use sumLong() as a direct replacement of consider more general purpose sumAll()
```
resolved in commit

kiwisincebirth commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -86,10 +286,57 @@ fun <T: Comparable<T>> Flux<T>.max(): Mono<T> = MathFlux.max(this)
+ * @author Mark Pruden
+ * @since 1.1.5
+ */
+fun <T> Flux<T>.sumInt(mapper: (T) -> Number): Mono<Int>
```
resolved

kiwisincebirth commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -115,11 +424,66 @@ fun <T> Flux<T>.sumDouble(mapper: (T) -> Number): Mono<Double>
+ * @since 1.0.0
+ * @deprecated Please use averageDouble() as a direct replacement of consider more general purpose averageAll()
  */
 fun <T> Flux<T>.average(mapper: (T) -> Number): Mono<Double>
```
resolved

kiwisincebirth commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -54,29 +177,106 @@ fun <T: Number> Flux<T>.sumDouble(): Mono<Double> = MathFlux.sumDouble(this)
+ * The resultant [Mono] will have the same [Number] type as the output of the mapping function
+ *
+ * If the result type (or precision) needs to be changed then a specific sumType() method
+ * should be used in preference to this method. e.g. sumDouble(mapper) can be used to compute
```
addressed

kiwisincebirth commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -115,11 +424,66 @@ fun <T> Flux<T>.sumDouble(mapper: (T) -> Number): Mono<Double>
  * @author Simon Baslé
- * @since 3.1.1
+ * @since 1.0.0
+ * @deprecated Please use averageDouble() as a direct replacement of consider more general purpose averageAll()
```
I have improved the documentation but kept averageAsDouble() as the replacement

kiwisincebirth commented:
First sorry for not resolving the tests before prior commit. 

> One last naming nitpick/proposal then: what do you think about sumAsXxx to better convey that a conversion is potentially in play? (sumAsLong, sumAsDouble, etc...)

Yes I like that have changed to include 'As' in all typed methods i.e. sumAsType(), averageAsType()

> oh and while I agree the sumType() methods have value, I really believe the sumType(Function) and averageType(Function) ones should be removed (since sumAll(Function)/averageAll(Function) would be expressive and concise enough to cover these cases).

Yep you are right on that, have removed the typed methods with function parameters, that I introduced, cannot remove sumDouble(Function) as it is part of the library

> missing @Deprecated annotation. I'm not entirely sure of the syntax and the annotation might need to be imported, as I'm suggesting this from Github's UI.

Added these annotations to the deprecated methods.

Have been through all the comments from above and marked resolved everything where I fixed. See response to your assumption above. 

Will commit shortly.

simonbasle commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -57,8 +58,11 @@ inline fun <reified T:Number> Flux<T>.sumAll(): Mono<T> =
  * @author Simon Baslé
  * @since 1.0.0
- * @deprecated Please use sumLong() as a direct replacement of consider more general purpose sumAll()
+ * @deprecated Please use sumLong() as a direct replacement, or consider more general purpose sumAll().
```
dang. it's now `sumAsLong` isn't it? deprecated messages need to be updated (doc, annotation message and annotation replaceWith) here and in the other similar deprecations

kiwisincebirth replied:
resolved

simonbasle commented on src/main/kotlin/reactor/kotlin/extra/math/MathFluxExtensions.kt:
```diff
@@ -178,8 +199,11 @@ inline fun <reified T:Number> Flux<T>.averageAll(): Mono<T> =
  * @author Simon Baslé
  * @since 1.0.0
- * @deprecated Please use averageDouble() as a direct replacement of consider more general purpose averageAll()
+ * @deprecated Please use averageDouble() as a direct replacement,
```
update method name here too

kiwisincebirth replied:
resolved

kiwisincebirth commented:
pleasure.
